src/styles/text/text.js

```diff
@@ -0,0 +1,371 @@
+        if (tile_data.queue.length > 0) {
+            tile = tile_data.queue[0][2].tile.key;
+            count = Object.keys(this.texts[tile]||{}).length;
+            console.log(`# texts for tile ${tile}: ${count}`);
+        }
+        if (!count) {
+            return Promise.resolve();
+        }
+
+        // Attach tile-specific label atlas to mesh as a texture uniform
+        tile_data.uniforms = { u_textures: ['labels-'+tile] };
+
+        // first call to main thread, ask for text pixel sizes
+        return WorkerBroker.postMessage('Text', 'getTextSizes', tile, this.texts[tile]).then(texts => {
+            this.bboxes[tile] = [];
+
+            // cleanup of texts that should be removed after occlusion test
+            for (let style in texts) {
+                let text_infos = texts[style];
+
+                for (let text in text_infos) {
+                    let text_info = text_infos[text];
+                    let label;
+                    let keep_in_tile;
+                    let move_in_tile;
+                    let geometry = this.geometries[tile][style][text];
+
+                    if (geometry.type === "LineString") {
+                        let exceed_heuristic = this.label_style.lines.exceed;
+                        let lines = geometry.coordinates;
+                        let line = [lines[0]];
+
+                        keep_in_tile = true;
+                        move_in_tile = true;
+
+                        label = new Label(text, line[0], text_info.size, lines);
+                    } else if (geometry.type === "Point") {
+                        let points = [geometry.coordinates];
+
+                        keep_in_tile = true;
+                        move_in_tile = false;
+
+                        label = new Label(text, points[0], text_info.size);
+                    }
+
+                    if (label.discard(move_in_tile, keep_in_tile, this.bboxes[tile], exceed_heuristic)) {
+                        // remove the text from the map
+                        delete text_infos[text];
+                    }
+
+                    text_info.label = label;
+                }
+            }
+
+            if (Object.keys(texts).length === 0) {
+                // early exit
+                return;
+            }
+
+            // second call to main thread, for rasterizing the set of texts
+            return WorkerBroker.postMessage('Text', 'addTexts', tile, texts).then(texts => {
+                this.texts[tile] = texts;
+
+                // Build queued features
+                tile_data.queue.forEach(q => this.super.addFeature.apply(this, q));
+                tile_data.queue = [];
+
+                return this.super.endData.call(this, tile_data);
+            });
+        });
+    },
+
+    // Override to queue features instead of processing immediately
+    addFeature (feature, rule, context, tile_data) {
+        // Collect text
+        let text = feature.properties.name;
```

Comment:
One thing I **would** like to do for v1 is to allow the field to use for the text to be specifiable in the stylesheet -- `name` was just an initial test, though it will probably be the most common value, and would be a good default. Following our typical style logic, I think a good behavior would be something like a `style.text_source` (name TBD, unsure what is best), that:

- defaults to feature `name` (`feature.properties.name`) when unset
- uses the value as the property name when set to a string, e.g. `style.text_source: value` would look for `feature.properties.value`
- can be set to a dynamic function, e.g. `style.text_source: function() { return (feature.value > 50 ? 'high' : 'low'); }` would set the text to either 'high' or 'low' depending on the feature.

@karimnaaji @meetar what do you think